File: src/parser/nim/utils/tilde.nim
import os, strutils

let hdir = getEnv("HOME")

# Expands starting tilde ('~') in path.
#
# @param {string} 1) - Path to expand.
# @return {string} - The expanded path.
#
# @resource: [https://nim-lang.org/docs/html#expandTilde%2Cstring]
proc expand*(p: string): string =
    runnableExamples:
        import os
        let hdir = getEnv("HOME")

        doAssert expand("") == ""
        doAssert expand("~") == hdir
        doAssert expand("~/Desktop") == joinPath(hdir, "Desktop")

    result = p
    if p.len == 0: result = ""
    elif p[0] == '~': result = hdir & p[1 .. p.high]

# Collapse starting home dir in a path to '~'.
#
# @param  {string} p - The path.
# @return {undefined} - Nothing is returned.
proc shrink*(p: string): string =
    runnableExamples:
        import os
        let hdir = getEnv("HOME")

        doAssert shrink("") == ""
        doAssert shrink(hdir) == "~"
        doAssert shrink(hdir & "/Desktop") == "~/Desktop"

    result = p
    if p.startsWith(hdir): result = "~" & p[hdir.len .. ^ 1]

File: src/parser/js2/utils/tilde.js
"use strict";

const path = require("path");
const hdir = require("os").homedir();

/**
 * Collapse starting home dir in a path to '~'.
 *
 * @param  {string} p - The path.
 * @return {undefined} - Nothing is returned.
 */
let shrink = (p) => {
	return p.startsWith(hdir) ? path.join("~", p.slice(hdir.length)) : p;
};

/**
 * Expand starting '~' in a path.
 *
 * @param  {string} p - The path.
 * @return {undefined} - Nothing is returned.
 */
let expand = (p) => (p.startsWith("~") ? path.join(hdir, p.slice(1)) : p);

module.exports = { shrink, expand };
